src/rtc.c

```diff
@@ -0,0 +1,518 @@
+ *
+ * The RTC sends and receives the date/time in bytes of binary-coded decimal
+ * numbers. This subsystem handles decoding and encoding the date/time into
+ * a more generally-useful struct called #rtc_time_t.
+ *
+ * Special thanks to marshallh and jago85 for their hard work and research
+ * reverse-engineering and documenting the inner-workings of the Joybus RTC.
+ *
+ * @{
+ */
+
+/**
+ * @brief Real-time clock identifier byte.
+ *
+ * A magic value in the RTC status response that indicates the RTC is present.
+ */
+#define JOYBUS_RTC_IDENTIFIER 0x10
+
+/**
+ * @brief Duration (in milliseconds) to wait after writing to RTC.
+ *
+ * The software must wait for the previous RTC write to finish before issuing
+ * another RTC write command. Since there is no status returned by the RTC to
+ * indicate that a write has completed, the software must wait a fixed duration.
+ */
+#define JOYBUS_RTC_WRITE_DELAY 20
+
+/**
+ * @brief The RTC is ready for block 2 to be read.
+ */
+#define JOYBUS_RTC_STATUS_READY 0x00
+/**
+ * @brief The RTC is ready for block 2 to be written.
+ */
+#define JOYBUS_RTC_STATUS_BUSY  0x80
+
+/**
+ * @brief Control mode for setting the RTC date/time.
+ *
+ * Clears the write-protection bits and sets the busy bit.
+ */
+#define JOYBUS_RTC_CONTROL_MODE_SET 0x0004
+/**
+ * @brief Control mode for normal operation of the RTC.
+ *
+ * Clears the busy bit sets the write-protection bits.
+ *
+ * 0x0100 is the block 1 write-protection bit.
+ * 0x0200 is the block 2 write-protection bit.
+ *
+ * RTC block 1 has an unknown purpose and is typically unimplemented.
+ * RTC block 2 contains the date and time.
+ */
+#define JOYBUS_RTC_CONTROL_MODE_RUN 0x0300
+
+/**
+ * @brief Decode a binary-coded decimal number.
+ *
+ * @param[in]   bcd
+ *              binary-coded decimal number
+ *
+ * @return the decoded number
+ */
+static uint8_t bcd_to_byte(uint8_t bcd)
+{
+    uint8_t hi = (bcd & 0xF0) >> 4;
+    uint8_t lo = bcd & 0x0F;
+    return (hi * 10) + lo;
+}
+
+/**
+ * @brief Encode a binary-coded decimal number.
+ *
+ * @param[in]   byte
+ *              number to encode
+ *
+ * @return the encoded binary-coded decimal number
+ */
+static uint8_t byte_to_bcd(uint8_t byte)
+{
+    byte %= 100;
+    return ((byte / 10) << 4) | (byte % 10);
+}
+
+/**
+ * @brief Compare two RTC times for (rough) equality
+ *
+ * @param[in]   a
+ *              Source pointer to an RTC time data structure
+ *
+ * @param[in]   b
+ *              Source pointer to an RTC time data structure
+ *
+ * @return whether the RTC time data structures are (roughly) equal
+ */
+static bool joybus_rtc_time_equal( const rtc_time_t * a, const rtc_time_t * b )
+{
+    return (
+        a->year  == b->year  &&
+        a->month == b->month &&
+        a->day   == b->day   &&
+        a->hour  == b->hour  &&
+        a->min   == b->min
+        /* Skip sec comparison in case the clock ticked between RTC reads */
```

Comment:
Technically, min could tick as well if sec was 59. Since this function is not very well defined semantically speaking, I'd inline it in its only caller, where ignoring seconds is OK because it is set as 3 (so it's unlikely to reach 60 before we read it... though multi threading might interfere)

Reply:
Addressed in https://github.com/DragonMinded/libdragon/pull/152/commits/e46fbb07a6cbb01a98f05a2d3b170d4ff1860b77